File: tts.js
#!/usr/bin/env node
# tiger 05/26/2019

var http = require('http'), 
  fs = require('fs'), 
  url = require('url');
const { exec } = require('child_process');
var tts = '~/bin/ekho.sh'

http.createServer(function(req, rsp) {
  var query = url.parse(req.url, true).query
  var txt = query.txt
  var p = query.p || 0
  var a = query.a || 0
  var s = query.s || 0
  var filename = Math.random().toString(36).substr(2)
  var file = '/tmp/'+filename

  if (! txt) {
    rsp.end('txt, Send request to TTS server \n'
	  + 'p, Set delta pitch. Value range from -100 to 100 (percent) \n'
	  + 'a, Set delta volume. Value range from -100 to 100 (percent) \n'
	  + 's, Set delta speed. Value range from -50 to 300 (percent)\n'
	)
	return
  }

  var cmd = tts+' -togg -p'+p+' -a'+a+' -s'+s+' -o'+file+' '+txt
  console.log(cmd)

  exec(cmd).on('exit', (code) => {
    rsp.writeHead(200, {
        'Content-Type': 'application/ogg',
        'Content-Length': fs.statSync(file).size
    })

    fs.createReadStream(file).pipe(rsp)
    fs.unlink(file, () => {})
  })

}).listen(2000);

File: tts.py
#!/usr/bin/env python2.7
#encode=utf-8

import random, os, re
import commands, urllib
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer

tts = '/bin/ekho.sh'
tts_help = '''Syntax: http://ip/?txt=
  txt, Send request to TTS server
  p, Set delta pitch. Value range from -100 to 100 (percent)
  a, Set delta volume. Value range from -100 to 100 (percent)
  s, Set delta speed. Value range from -50 to 300 (percent)
'''

class MyHandler(BaseHTTPRequestHandler):
 
  def getparam(self, name):
    m = re.match(r'.*'+name+'=([^&]*)', self.path)
    return m and m.group(1) or '0'

  def do_GET(self):
    txt = self.getparam('txt')
    if txt == '0':
      self.send_response(200)
      self.send_header('Content-Type', 'text/plain')
      self.end_headers()
      self.wfile.write(tts_help)
      return

    p = self.getparam('p')
    a = self.getparam('a')
    s = self.getparam('s')

    mfile='/tmp/tts-'+str(random.randint(10000,99999))
    cmd = tts+' -togg -p'+p+' -a'+a+' -s'+s+' -o'+mfile+' '+urllib.unquote(txt)
    print cmd
    (status, output) = commands.getstatusoutput(cmd)

    f = open(mfile, 'rb')
    self.send_response(200)
    self.send_header('Content-Type', 'application/ogg')
    self.end_headers()
    self.wfile.write(f.read())
    f.close()

    os.unlink(mfile)

def main():
  try:
    server = HTTPServer(('0.0.0.0', 8000), MyHandler)
    server.serve_forever()
  except KeyboardInterrupt:
    server.socket.close()

if __name__ == '__main__':
  main()
